src/posts/lucky_0.25.cr

```diff
@@ -0,0 +1,370 @@
+
+    Prior to this update, it was sort of "wild west", and completely up to you on how to implement these. For consistency sake, we recommended defining a `submit`
+    method, and then returning `yield self, value`. These operations were also limited with no way to use callbacks, file attributes, or define errors not tied to
+    an attribute.
+
+    In this release, we've created a whole new interface!
+
+    ```crystal
+    class PromoteUserValidator < Avram::Operation
+      param_key :user
+      needs user : User
+      attribute token : String
+
+      # before callbacks
+      before_run do
+        if user.has_inactive_account?
+          # custom errors
+          add_error(:user_inactive, "My custom error message")
+        end
+
+        validate_required(token)
+      end
+
+      # after callbacks
+      after_run do |user|
+        NotifyUser.new(user.email).deliver
+      end
+
+      # unified interface
+      def run
+        if user.is_promotable?
+          user
+        else
+          nil
+        end
+      end
+    end
+
+    PromoteUserValidator.run(params) do |operation, user|
+      if user
+        # the user must be promotable
+      else
+        # no user, lets check our custom error
+        operation.errors[:user_inactive]
+      end
+    end
+    ```
+
+    ### Better callbacks in `Avram::SaveOperation`
+
+    We've also made quite a few changes to the `Avram::SaveOperation` family as well.
+
+    Prior to this update, your `after_save` and `after_commit` callbacks couldn't be used with an anonymous block like how `before_save` works. Well,
+    now they can!
+
+    ```crystal
+    class SaveUser < User::SaveOperation
+      before_save do
+        # before save
+      end
+
+      after_save do |saved_user|
+        # after save
+      end
+
+      after_commit do |saved_user|
+        # after DB commit
+      end
+    end
+    ```
+
+    We have also added the ability to do these callbacks conditionally. This allows you to specify if the callbacks should actually run or not.
+
+    ```crystal
+    class SaveUser < User::SaveOperation
+      before_save :validate_card_number, if: :new_transaction?
+
+      after_save :update_billing, unless: :card_still_valid?
+
+      private def validate_card_number
+      end
+
+      private def new_transaction?
+        true
+      end
+
+      private def update_billing
+      end
+
+      private def card_still_valid?
+        false
+      end
+    end
+    ```
+
+    Lastly, there were some instances where you may be updating a record, and expecting a callback when everything was done. However, in the case of
+    a record that doesn't need to update anything, you'd never have any callbacks that fired. (i.e. kick off background job, or email, etc...)
+
+    To help with this, we've added 1 additional callback called `after_completed`. This callback is always called when the `save_status` is set to `:saved`.
```

Comment:
Seems like this comment on `save_status` is talking about internal implementation and not necessary to understand how `after_completed` works.

```markdown
To help with this, we've added an additional callback called `after_completed`.
It will always run after valid operations have completed, even if there is nothing to save to the database.
```